### ISSUE TITLE
Containers with calico interfaces can't communicate with control node

### ISSUE THREAD

**Issue by dpavlos** [mantl/mantl#1944]
I noticed this strange issue while I was trying to debug this one https://github.com/mantl/mantl/issues/1799 .

As far as I can see there are two calico profiles configured on k8s cluster: `dev` and `ng_calico-k8s-network` . For the `ng_calico-k8s-network` profile there are two endpoints created on `kubeworker-001` with the following workloads assigned:

`kube-system.kube-dns-v8-v8zfk`
`kube-system.kubernetes-dashboard-manj0`

On kubeworker-001:

```
# calicoctl endpoint show --detail
+----------------+-----------------+----------------------------------------+----------------------------------+----------------+-------------------+-----------------------+--------+
|    Hostname    | Orchestrator ID |              Workload ID               |           Endpoint ID            |   Addresses    |        MAC        |        Profiles       | State  |
+----------------+-----------------+----------------------------------------+----------------------------------+----------------+-------------------+-----------------------+--------+
| kubeworker-001 |       k8s       |     kube-system.kube-dns-v8-v8zfk      | 22e032f0280f11e7a126080027a95b09 | 192.168.0.2/32 | 3e:cd:4a:ae:f0:d4 | ng_calico-k8s-network | active |
| kubeworker-001 |       k8s       | kube-system.kubernetes-dashboard-manj0 | 22aa20e8280f11e7978b080027a95b09 | 192.168.0.1/32 | 0a:6e:d0:b8:fc:83 | ng_calico-k8s-network | active |
+----------------+-----------------+----------------------------------------+----------------------------------+----------------+-------------------+-----------------------+--------+
```

```
# calicoctl profile show --detail
+-----------------------+----------------+-----------------+----------------------------------------+----------------------------------+--------+
|          Name         |      Host      | Orchestrator ID |              Workload ID               |           Endpoint ID            | State  |
+-----------------------+----------------+-----------------+----------------------------------------+----------------------------------+--------+
|          dev          |      None      |       None      |                  None                  |               None               |  None  |
| ng_calico-k8s-network | kubeworker-001 |       k8s       |     kube-system.kube-dns-v8-v8zfk      | 22e032f0280f11e7a126080027a95b09 | active |
| ng_calico-k8s-network | kubeworker-001 |       k8s       | kube-system.kubernetes-dashboard-manj0 | 22aa20e8280f11e7978b080027a95b09 | active |
+-----------------------+----------------+-----------------+----------------------------------------+----------------------------------+--------+
```

From the above it seems that both `kube2sky` and `kubernetes-dashboard` containers are using calico internal IPs from the calico pool

```
# calicoctl pool show
+----------------+-------------------+
|   IPv4 CIDR    |      Options      |
+----------------+-------------------+
| 192.168.0.0/16 | ipip,nat-outgoing |
+----------------+-------------------+
+--------------------------+---------+
|        IPv6 CIDR         | Options |
+--------------------------+---------+
| fd80:24e2:f998:72d6::/64 |         |
+--------------------------+---------+
```

Tried to ping `control-01` from `kubeworker-001/kube2sky` container and even if it seems that there is a reply from `control-01` nothing reaches the container.

tcpdump on control-01
```
# tcpdump -n -i enp0s8 -p icmp
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on enp0s8, link-type EN10MB (Ethernet), capture size 65535 bytes
11:24:52.317014 IP 172.20.10.151 > 172.20.10.101: ICMP echo request, id 3584, seq 0, length 64
11:24:52.317083 IP 172.20.10.101 > 172.20.10.151: ICMP echo reply, id 3584, seq 0, length 64
```

tcpdump on kubeworker-001
```
# tcpdump -n -i enp0s8 icmp
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on enp0s8, link-type EN10MB (Ethernet), capture size 65535 bytes
11:24:52.316707 IP 172.20.10.151 > 172.20.10.101: ICMP echo request, id 3584, seq 0, length 64
11:24:52.317733 IP 172.20.10.101 > 172.20.10.151: ICMP echo reply, id 3584, seq 0, length 64
```

On `control-01` node I get the following error in the kube-dns pod:

```
$ kubectl logs kube-dns-v8-v8zfk -c kube2sky
I0423 10:55:45.148082       1 kube2sky.go:383] [Attempt: 1] Attempting access to etcd after 5 second sleep
I0423 10:56:20.149793       1 kube2sky.go:383] [Attempt: 2] Attempting access to etcd after 5 second sleep
I0423 10:56:55.150713       1 kube2sky.go:383] [Attempt: 3] Attempting access to etcd after 5 second sleep
I0423 10:57:30.153255       1 kube2sky.go:383] [Attempt: 4] Attempting access to etcd after 5 second sleep
I0423 10:58:05.155058       1 kube2sky.go:383] [Attempt: 5] Attempting access to etcd after 5 second sleep
I0423 10:58:40.165374       1 kube2sky.go:383] [Attempt: 6] Attempting access to etcd after 5 second sleep
I0423 10:59:15.167657       1 kube2sky.go:383] [Attempt: 7] Attempting access to etcd after 5 second sleep
I0423 10:59:50.168211       1 kube2sky.go:383] [Attempt: 8] Attempting access to etcd after 5 second sleep
I0423 11:00:25.168965       1 kube2sky.go:383] [Attempt: 9] Attempting access to etcd after 5 second sleep
I0423 11:01:00.175460       1 kube2sky.go:383] [Attempt: 10] Attempting access to etcd after 5 second sleep
I0423 11:01:35.177550       1 kube2sky.go:383] [Attempt: 11] Attempting access to etcd after 5 second sleep
F0423 11:02:10.178910       1 kube2sky.go:512] Failed to create etcd client - failed to connect to etcd server: http://172.20.10.101:2379, error: Get http://172.20.10.101:2379/version: dial tcp 172.20.10.101:2379: i/o timeout
```

The i/o timeout error during the connection attempt with the control node (172.20.10.101) is exactly the same error we have with the kubernetes-dashboard. Perhaps this is related also with https://github.com/mantl/mantl/issues/1942

**Comment by KaGeN101** [comment]
This is related to #1942

**Comment by KaGeN101** [comment]
192.168.0.0/16 is the calico container network and there is definitely an issue with getting traffic to go outside the cluster

**Comment by KaGeN101** [comment]
Perhaps just adding a route from 192.168.0.0/16 to 10.0.0.0/8 will solve this

**Comment by KaGeN101** [comment]
Got both dns and and dashboard working by placing then on the hostNetwork. So it is not using calico at all now for any of the kube-system pods.

Add this patch as a work around but is not the solution if we want to use calico will have to add a route that forward calico traffic to cluster network, have not idea how to do this but will also work:

diff --git a/roles/kubernetes-master/templates/kubernetes-dashboard.yml b/roles/kubernetes-master/templates/kubernetes-dashboard.yml
index 845541c..55bd754 100644
--- a/roles/kubernetes-master/templates/kubernetes-dashboard.yml
+++ b/roles/kubernetes-master/templates/kubernetes-dashboard.yml
@@ -32,6 +32,7 @@ items:
         labels:
           app: kubernetes-dashboard
       spec:
+        **hostNetwork: true**
         containers:
         - name: kubernetes-dashboard
           image: gcr.io/google_containers/kubernetes-dashboard-amd64:v1.1.0
diff --git a/roles/kubernetes-master/templates/skydns-rc.yaml.j2 b/roles/kubernetes-master/templates/skydns-rc.yaml.j2
index 2ee5319..9434807 100644
--- a/roles/kubernetes-master/templates/skydns-rc.yaml.j2
+++ b/roles/kubernetes-master/templates/skydns-rc.yaml.j2
@@ -19,6 +19,7 @@ spec:
         version: v8
         kubernetes.io/cluster-service: "true"
     spec:
+      **hostNetwork: true**
       containers:
       - name: kube2sky
         image: gcr.io/google_containers/kube2sky:1.11

**Comment by KaGeN101** [comment]
We should decide if this is the fix or add the routing for the calico network(which I dont know how to do :( )

**Comment by dpavlos** [comment]
https://github.com/mantl/mantl/pull/1945

**Comment by Theaxiom** [comment]
Okay, I've parked this workaround in a branch: https://github.com/mantl/mantl/tree/k8-workaround

In the meantime I am going to work on Calico and getting that working properly.

**Comment by Theaxiom** [comment]
This may be resolved by #1941

**Comment by dpavlos** [comment]
Fixed in master https://github.com/mantl/mantl/pull/1963. Closing the issue.
